pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -39,284 +108,413 @@ func Add(mgr manager.Manager) error {
-	// This client, initialized using mgr.Client() above, is a split client
-	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
-	podInformer cache.SharedInformer
-}
+		case LeaseStatusActiveOwnedByDifferentNode:
+			{
+				if statusAnnoData.status == NodeStateInit {
+					r.reqLogger.Infof("wait to acquire lease. requeue. node: %s\n", request.Name)
+					r.setStatusAnnotation(node, statusAnnoData)
+					return reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+				}
 
-var Handler ReconcileHandler
+				statusAnnoData.status = NodeStateLeaseNotOwnedByNode
+				r.setStatusAnnotation(node, statusAnnoData)
 
-// Reconcile reads that state of the cluster for a NodeMaintenance object and makes changes based on the state read
-// and what is in the NodeMaintenance.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	reqLogger := log.WithFields(log.Fields{"Request.Namespace": request.Namespace, "Request.Name": request.Name})
-	reqLogger.Info("Reconciling NodeMaintenance")
+				err := fmt.Errorf("lease got acquired by someone while request was active. node: %s", request.Name)
+				return reconcile.Result{}, err
+			}
 
-	// Fetch the NodeMaintenance instance
-	instance := &kubevirtv1alpha1.NodeMaintenance{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			// Request object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			reqLogger.Infof("NodeMaintenance Object: %s Deleted ", request.NamespacedName)
-			return reconcile.Result{}, nil
+		case LeaseStatusFail:
+			{
+				r.reqLogger.Errorf("failed to get lease object. node: %s error: %v", request, err)
+
+				statusAnnoData.status = NodeStateFailedToGetLease
+				r.setStatusAnnotation(node, statusAnnoData)
+
+				return reconcile.Result{}, err
+			}
 		}
-		// Error reading the object - requeue the request.
-		reqLogger.Info("Error reading the request object, requeuing.")
-		return reconcile.Result{}, err
-	}
 
-	// Add finalizer when object is created
-	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
-		if !ContainsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
-			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer)
-			if err := r.client.Update(context.TODO(), instance); err != nil {
-				return r.reconcileAndError(instance, err)
+		if drainer, kclient, err := r.initDrainer(); err == nil {
```

Comment:
Why create this each time?

Reply:
good question - I didn't know of a simple way to maintain additional state with the reconcile request. Do you have an idea? (actually the PR you sited creates the drainer again and again)

Reply:
Perhaps add it to `r` ?

Reply:
r is ReconcileNodeMaintenance - we have one such object per control loop; therefore we would need to add  a map from Request object to some structure that holds the drainer and delete that one when the request is completed. Do you approve of this course of action?

Reply:
Are you thinking that there can be multiple calls to `Reconcile()` at the same time?